### PR TITLE
Allocation free iterators for LinkedHashMap

### LINKED ISSUE #3486: LinkedHashMap creates enormous amount of memory traffic
Because of an efficient implementation of Enumerable pattern.
It creates a significant pressure to GC.

Most of the traffic come from boxed `KeyValuePair<CollectionEntry, IPersistentCollection>` values, but Enumerator also could be made a `struct` for removing allocations further.

During a startup our application caches a lot of data.
So, the image below is for our case.
image

I work on a fix, check my solution against original using DotNetBenchmark.

### PR DESCRIPTION
**Pull request by wilbit**
fixes #3486

### CODE REVIEW

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkedHashMap.cs:185]: Such changes are binary breaking, the class and property being public. So, such a change could only be merged for a new major version. For a faster release of the change, you need to ensure all changes are non-breaking. So, you would have to introduce new properties/methods and obsolete the old ones, instead of changing in a breaking way existing ones.

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkedHashMap.cs:403]: To avoid increasing the API for which we have to avoid breaking changes, better switch it to internal instead, if possible.

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkedHashMap.cs:480]: Better make it internal if possible.

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkedHashMap.cs:528]: Same here.

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkedHashMap.cs:605]: Same here.

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkedHashMap.cs:653]: Same here.

File: src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs:412]: `.GetType` will return a real type of instance, I needed to check a type of variable (basically, a return type of chosen `.GetEnumerator`)

File: src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs:182]: I wanted to check that it is struct, so, I would prefer to keep 1 assertion (`Type` does not have `IsStruct` property =()

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs:185]: According to your https://github.com/nhibernate/nhibernate-core/issues/2407#issuecomment-636494042

> NHibernate is not meant to be an utility library. This kind of classes should have been made internal in my opinion. In other words, we do not support external uses of LinkedHashMap.

So, maybe, it is even better to make `LinkedHashMap` class internal.

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs:403]: I cannot make it internal because it is used in a public method.
I would suggest to make `LinkedHashMap` internal.

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs:480]: same here

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs:528]: same here

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs:605]: same here

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs:653]: same here

File: src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs:192]: same here

File: src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs:202]:
```suggestion
			Assert.That(enumeratorType.IsClass || enumeratorType.IsInterface, Is.False, "The enumerator should be a struct");
```

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs:185]: I cannot use `ObsoleteAttribute` here, because a class cannot have 2 properties with the same name.

File: src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs
Comment (fredericDelaporte) [reply src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs:202]: ok

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [reply src/NHibernate/Util/LinkedHashMap.cs:185]: > it is even better to make LinkedHashMap class internal

Yes it is better, but that would be a breaking change, so, it would have to wait for a major release.

> I don't care about getting these changes faster.

Unless we start issuing major versions for rather minor changes, I do not think a major version will come any time soon. There are many things that should be done for an actual new major version, but I currently do not see enough contributing workforce available to do them. There are approved PR in the next major milestone lingering there since years...

> I cannot use ObsoleteAttribute here, because a class cannot have 2 properties with the same name.

But you can add it with another name. Granted, that would not be ideal.

I know `GetEnumerator` has a special treatment by the compiler for foreach statement, so for this one, using another name is not an option.

**Comment by hazzik** [comment]
@wilbit @fredericDelaporte would it be feasible to create a new class and use it?

**Comment by wilbit** [comment]
> @wilbit @fredericDelaporte would it be feasible to create a new class and use it?

Well, I see that `LinkedHashMap` is not used in public properties/fields/method, so, technically it is feasible.
But also, `LinkedHashMap` implements `IDeserializationCallback` interface, and I'm not really sure what it brings. If nothing, that we are good to go and the only thing which is left is a new name and/or namespace for the internal class.

Another suggestion from me:
1. keep the changes from this PR for a next major release (after some fixes and a rebase, see point 2)
2. for an upcoming minor release fix the major part of #3486 by casting `LinkedHashMap` to `IDictionary<TKey, TValue>` in `BatchFetchQueue.GetCollectionBatchAsync` (this should do a trick and `foreach` will use generic `IEnumerator<>` instead of non-generic `IEnumerator`).

**Comment by fredericDelaporte** [comment]
Obsoleting LinkedHashMap entirely and adding an internal replacement seems to me as the best way to handle this.

> But also, `LinkedHashMap` implements `IDeserializationCallback` interface, and I'm not really sure what it brings.

It is for supporting binary serialization. See also the `Serializable` attribute of the class. That thing is obsoleted in .Net Core for security reasons, but it is still used by many users especially under .Net Framework. The replacement class will have to be binary serializable too, because it will serve as state of many other classes which are serializable too. This allows to serialize the whole session and restore it later, by example.

**Comment by wilbit** [comment]
@fredericDelaporte
1. OK, where to put the new internal class and how to name it?
2. about serialisation: would it work if somebody serialize a session with the current `LinkedHashMap` and deserialize with internal `LinkedHashMap` (which is, basically, would be a copy of the current `LinkedHashMap` in terms of fields...)?

**Comment by fredericDelaporte** [comment]
1. Why not `LinkHashMap`, same place?
2. That will not work and will be a possible breaking change, but neither a binary one nor a source one, which is acceptable for a minor. We will have to warn in the release notes about the incompatibility of states serialized by version prior to the next minor with this next minor, and vice-versa.

**Comment by wilbit** [comment]
added LinkHashMap, @hazzik , @fredericDelaporte

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkedHashMap.cs:23]: It will be 5.6. No need to mention the replacement since it will not be available for external use.

```suggestion
	// Since 5.6
	[Obsolete("This class has no more usages and will be removed in a future version.")]
```

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:30]: To avoid the `this`, switch to `_` prefix for class members.

```suggestion
			private readonly TKey _key;
			private TValue _value;
			private Entry _next;
			private Entry _prev;
```

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:88]:
```suggestion
		private readonly Entry _header;
		private readonly Dictionary<TKey, Entry> _entries;
		private long _version;
```

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:385]:
```suggestion
			private readonly LinkHashMap<TKey, TValue> _dictionary;
```

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:449]:
```suggestion
				private readonly LinkHashMap<TKey, TValue> _dictionary;
				private Entry _current;
				private readonly long _version;
```

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:495]:
```suggestion
			private readonly LinkHashMap<TKey, TValue> _dictionary;
```

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:559]:
```suggestion
				private readonly LinkHashMap<TKey, TValue> _dictionary;
				private Entry _current;
				private readonly long _version;
```

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:607]:
```suggestion
			private readonly LinkHashMap<TKey, TValue> _dictionary;
			private Entry _current;
			private readonly long _version;
```

File: src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs:139]: Got it...

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs:30]: I thought it was o code style :)))

File: src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkedHashMapFixture.cs]: applied

File: src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs]: applied `Assert.That(entries, Is.Zero, "should not have any entries in the enumerators");`

File: src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs]: applied

File: src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs]: applied

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: applied

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: applied

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: applied

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: applied

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: applied

File: src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs]: applied

File: src/NHibernate/Util/LinkedHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkedHashMap.cs]: applied

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: applied

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: applied

File: src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs
Comment (wilbit) [reply src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs:91]: I do not change that because:
- The suggestion does not work, because `Does.Contain` checks existence against KeyValuePairs collection.
- `LinkHashMap.Contains(string)` is a target of this test.

Comment (fredericDelaporte) [reply src/NHibernate.Test/UtilityTest/LinkHashMapFixture.cs:91]: Ok.

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [reply src/NHibernate/Util/LinkHashMap.cs]: There are no [Microsoft guidelines](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-fields) for private fields. `camelCase` naming without `_` prefix in NHibernate is generally the result of port from Java without renaming. This causes the fields to be looking like local variables and methods parameters. The `_` prefix is deemed ugly by some, but has the merit to allow obvious recognition of fields versus local variables and method parameters.

File: src/NHibernate/Util/LinkHashMap.cs
Comment (fredericDelaporte) [inline src/NHibernate/Util/LinkHashMap.cs:28]: The `e` was very likely a "trick" to dodge the fact that `value` is a reserved word. (Instead of using `@` which is meant for this purpose.) There are no reason to keep it.

```suggestion
			private TValue _value;
```

Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs:28]: Maybe, I wasn't sure because of the serialisation...
Also one question bother me: may I
- remove fields by using auto properties
- or remove properties and make fields "public"?
what would be a safe move to clean `Entry` class?

Comment (fredericDelaporte) [reply src/NHibernate/Util/LinkHashMap.cs:28]: You can remove fields by using auto properties.

Comment (wilbit) [reply src/NHibernate/Util/LinkHashMap.cs]: OK, I've renamed `_evalue` to `_value`, but I did not use auto properties because generated backend fields have longer names and I realised that size of binary serialized object (stream/array) depends on:
- length of field names, 1 byte per character
- length of parent class name for nested classes, 3 bytes per character
- length of class name, 3 bytes per character
- length of namespace name, 3 bytes per character

Current size of 1 serialized object like `LinkHashMap<int,int>.Entry` is 826 bytes long.
- Using of auto properties makes it 890 bytes long (+7.75%)
- Extracting `Entry` class directly to NHibernate namespace and using field names like:
```c#
private readonly TKey k;
private TValue v;
private Entry<TKey,TValue> n;
private Entry<TKey,TValue> p;
```
makes it 759 bytes long (-8.11%)

So, I would not use auto properties and also consider to rename fields to 1 character long names at least. This will make it 810 bytes long (-1.94%).

Any thought? Because I don't really know how important it is (I don't use serialisation in NHibernate and not familiar with this feature, but it seems important to mention).

Comment (fredericDelaporte) [reply src/NHibernate/Util/LinkHashMap.cs]: We do not really care about the resulting size of the serialized binary. I have no recollection of this subject having ever been raised. Binary serialization is for us a legacy that we do not want to break in a minor: as long as it keeps working, that is good enough for us.

So, no need to bother about the resulting size. Code cleanliness is more important for us than the impacts it may have on the size of state binary serialized. (No one letter backing fields please!)

**Review by fredericDelaporte — DISMISSED** [review]

**Comment by fredericDelaporte** [comment]
Master branch protection requires the PR to be up-to-date with master. This PR has not edits for maintainers enabled, so, we cannot update it ourselves. Can you update it?

File: src/NHibernate/Util/LinkHashMap.cs
Comment (wilbit) [inline src/NHibernate/Util/LinkHashMap.cs:45]: I've removed using of `Value` from `GetHashCode`

**Review by fredericDelaporte — COMMENTED** [review]
Master branch protection requires the PR to be up-to-date with master. This PR has not edits for maintainers enabled, so, we cannot update it ourselves. Can you update the PR branch or enable edits for maintainers?
If you update it without enabling edits for maintainers, please also regenerate async files, as the automated task for this fails without this permission.

**Comment by wilbit** [comment]
rebased, @fredericDelaporte

**Review by fredericDelaporte — APPROVED** [review]

**Pull request merged**
